File: NRDB/CardType.m
//
//  CardType.m
//  NRDB
//
//  Created by Gereon Steffens on 15.12.13.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import "CardType.h"
#import "CardManager.h"
#import "TableData.h"

@implementation CardType

static NSDictionary* code2type;
static NSMutableDictionary* type2name;
static NSMutableArray* runnerTypes;
static NSMutableArray* corpTypes;
static TableData* allTypes;

+(void) initialize
{
    code2type = @{
        @"identity": @(NRCardTypeIdentity),
        @"asset": @(NRCardTypeAsset),
        @"agenda": @(NRCardTypeAgenda),
        @"ice": @(NRCardTypeIce),
        @"upgrade": @(NRCardTypeUpgrade),
        @"operation": @(NRCardTypeOperation),
        @"program": @(NRCardTypeProgram),
        @"hardware": @(NRCardTypeHardware),
        @"resource": @(NRCardTypeResource),
        @"event": @(NRCardTypeEvent),
    };

    type2name = [NSMutableDictionary dictionary];
}
    
+(void) initializeCardTypes:(NSArray*)cards
{
    [type2name removeAllObjects];
    [type2name setObject:kANY forKey:@(NRCardTypeNone)];
    for (Card* c in cards)
    {
        [type2name setObject:c.typeStr forKey:@(c.type)];
        
        if (type2name.count == code2type.count + 1)
        {
            break;
        }
    }
    
    NRCardType rt[] = { NRCardTypeEvent, NRCardTypeHardware, NRCardTypeResource, NRCardTypeProgram };
    NRCardType ct[] = { NRCardTypeAgenda, NRCardTypeAsset, NRCardTypeUpgrade, NRCardTypeOperation, NRCardTypeIce };
    
    runnerTypes = [NSMutableArray array];
    for (int i=0; i<DIM(rt); ++i)
    {
        [runnerTypes addObject:[CardType name:rt[i]]];
    }
    
    corpTypes = [NSMutableArray array];
    for (int i=0; i<DIM(ct); ++i)
    {
        [corpTypes addObject:[CardType name:ct[i]]];
    }
    
    NSArray* typeSections = @[ @"", l10n(@"Runner"), l10n(@"Corp") ];
    NSMutableArray* types = [NSMutableArray array];
    [types addObject:@[ [CardType name:NRCardTypeNone ], [CardType name:NRCardTypeIdentity] ]];
    [types addObject:runnerTypes];
    [types addObject:corpTypes];
    
    allTypes = [[TableData alloc] initWithSections:typeSections andValues:types];

    [runnerTypes insertObject:[CardType name:NRCardTypeNone] atIndex:0];
    [corpTypes insertObject:[CardType name:NRCardTypeNone] atIndex:0];
}

+(NRCardType) type:(NSString*)code
{
    return [[code2type objectForKey:code] intValue];
}

+(NSString*) name:(NRCardType)type
{
    return [type2name objectForKey:@(type)];
}

+(NSArray*) typesForRole:(NRRole)role
{
    NSAssert(role != NRRoleNone, @"no role");
    return role == NRRoleRunner ? runnerTypes : corpTypes;
}

+(TableData*) allTypes
{
    return allTypes;
}

@end

File: NetDeck/CardType.swift
//
//  CardType.swift
//  NetDeck
//
//  Created by Gereon Steffens on 14.11.15.
//  Copyright © 2017 Gereon Steffens. All rights reserved.
//

import Foundation

// @objc to make NSPredicates on card.type work

@objc enum CardType: Int {
    
    case none = -1
    case identity
        
    // corp
    case agenda, asset, upgrade, operation, ice
        
    // runner
    case event, hardware, resource, program
    
    // NB: Card diff depends on ICE/Program being the last entries!
    private static let runnerTypes: [CardType] = [ .event, .hardware, .resource, .program ]
    private static let corpTypes: [CardType] = [ .agenda, .asset, .upgrade, .operation, .ice ]
    
    private static var type2name = [CardType: String]()
    static var runnerTypeNames = [String]()
    static var corpTypeNames = [String]()
    private(set) static var allTypes: TableData<String>!
    
    static func initializeCardType(_ typesDict: [CardType: String]) -> Bool {
        runnerTypeNames = []
        corpTypeNames = []
        
        type2name.removeAll()
        type2name[.none] = Constant.kANY
        type2name.merge(typesDict) { (current, _) in return current }
        for (type, str) in typesDict {
            type2name[type] = str
        }
        let expectedTypes = 11 // 10 card types, +1 for "Any"
        assert(type2name.count == expectedTypes)
        if type2name.count != expectedTypes {
            return false
        }
        
        for type in runnerTypes {
            runnerTypeNames.append(CardType.name(for: type))
        }
        for type in corpTypes {
            corpTypeNames.append(CardType.name(for: type))
        }
        
        let typeSections = [ "", "Runner".localized(), "Corp".localized() ]
        let types = [
            [CardType.name(for: .none), CardType.name(for: .identity)],
            runnerTypeNames,
            corpTypeNames
        ]
        
        allTypes = TableData(sections: typeSections, values: types)
        
        runnerTypeNames.insert(Constant.kANY, at: 0)
        corpTypeNames.insert(Constant.kANY, at: 0)
        
        return true
    }
    
    static func name(for type: CardType) -> String {
        return type2name[type] ?? "n/a"
    }
    
    static func typesFor(role: Role) -> [String] {
        assert(role != .none, "no role")
        return role == .runner ? runnerTypeNames : corpTypeNames
    }
}
